File: 09/ruby/max.rb
# max.rb tests a Ruby linked list.
#
# Begun by: Dr. Adams for CS 214 at Calvin College.
# Completed by: Charles Blum
# Date: 21 April 2014

require 'List'

list1 = List.new
list2 = List.new
list3 = List.new
list4 = List.new

list1.append(99).append(88).append(77).append(66).append(55)
list2.append(55).append(66).append(77).append(88).append(99)
list3.append(55).append(77).append(99).append(88).append(66)
list4.append(55).append(44).append(33).append(22).append(11)

list1.print 
puts
puts "Maximum of list1: #{list1.max}"
puts "The index of 99 is: #{list1.getIndex(99)}"

list2.print 
puts
puts "Maximum of list2: #{list2.max}"
puts "The index of 99 is: #{list2.getIndex(99)}"

list3.print 
puts
puts "Maximum of list3: #{list3.max}"
puts "The index of 99 is: #{list3.getIndex(99)}"

list4.print 
puts
puts "Maximum of list43: #{list4.max}"
puts "The index of 99 is: #{list4.getIndex(99)}"

File: 09/c++/max.cpp
/* max.cpp finds the maximum value in a C++ linked list.
 *
 * Begun by: Dr. Adams, CS 214 at Calvin College.
 * Completed by:Charles Blum
 * Date:21 April 2014
 */

#include <iostream>       // cin, cout, <<, >>
#include <algorithm> 	  // max_element
using namespace std;

#include "mylist.h"       // print()

int main() {
  // define list1, list2, list3, list4 ...
  list<int> list1, list2, list3, list4;

  list1.push_back(99);    // 99, 88, 77, 66, 55
  list1.push_back(88);    // max is first
  list1.push_back(77);
  list1.push_back(66);  
  list1.push_back(55);

  list2.push_back(55);    // 55, 66, 77, 88, 99
  list2.push_back(66);    // max is last
  list2.push_back(77);  
  list2.push_back(88);  
  list2.push_back(99);  
  
  list3.push_back(55);    // 55, 77, 99, 88, 66
  list3.push_back(77);    // max is in the middle
  list3.push_back(99);  
  list3.push_back(88);  
  list3.push_back(66);  
  
  print(list1, cout); cout << endl;
  print(list2, cout); cout << endl;
  print(list3, cout); cout << endl;

  // display maxima of the 3 lists...
  cout << *max_element(list1.begin(), list1.end()) << endl;
  cout << *max_element(list2.begin(), list2.end()) << endl;
  cout << *max_element(list3.begin(), list3.end()) << endl;
  
  // find the index of 99 in each of the lists.
  list4.push_back(55);    // 55, 44, 33, 22, 11
  list4.push_back(44);    // 99 is not present
  list4.push_back(33);  
  list4.push_back(22);  
  list4.push_back(11);  
  
  cout << "The index of 99 is: " << index(list1, 99) << endl;
  cout << "The index of 99 is: " << index(list2, 99) << endl;
  cout << "The index of 99 is: " << index(list3, 99) << endl;
  cout << "The index of 99 is: " << index(list4, 99) << endl;
}

